src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,3349 @@
+  return irep_idt(value);
+}
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::add_axioms_for_constant
+
+  Inputs: a string constant
+
+ Outputs: a string expression
+
+ Purpose: add axioms saying the returned string expression should be equal
+          to the string constant
+
+\*******************************************************************/
+
+string_exprt string_constraint_generatort::add_axioms_for_constant(
+  irep_idt sval)
+{
+  string_exprt res(get_char_type());
+  std::string c_str=sval.c_str();
+  std::wstring str;
+  // We should have a special treatment for java strings when the conversion
+  // function is available:
+  // if(mode==ID_java)
+  //   str=utf8_to_utf16_little_endian(c_str);
+  // else
+    str=widen(c_str);
+
+  for(std::size_t i=0; i<str.size(); i++)
+  {
+    exprt idx=from_integer(i, get_index_type());
+    exprt c=from_integer(str[i], get_char_type());
+    equal_exprt lemma(res[idx], c);
+    axioms.push_back(lemma);
+  }
+
+  exprt s_length=from_integer(str.size(), get_index_type());
+
+  axioms.push_back(res.axiom_for_has_length(s_length));
+  return res;
+}
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::add_axioms_for_empty_string
+
+  Inputs: function application without argument
+
+ Outputs: string expression
+
+ Purpose: add axioms to say that the returned string expression is empty
+
+\*******************************************************************/
+
+string_exprt string_constraint_generatort::add_axioms_for_empty_string
+(const function_application_exprt &f)
+{
+  assert(f.arguments().empty());
+  string_exprt res(get_char_type());
+  axioms.push_back(res.axiom_for_has_length(0));
+  return res;
+}
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::add_axioms_from_literal
+
+  Inputs: function application with an argument which is a string literal
+
+ Outputs: string expression
+
+ Purpose: add axioms to say that the returned string expression is equal to
+          the string literal
+
+\*******************************************************************/
+
+string_exprt string_constraint_generatort::add_axioms_from_literal(
+  const function_application_exprt &f)
+{
+  const function_application_exprt::argumentst &args=f.arguments();
+  assert(args.size()==1); // Bad args to string literal?
+
+  const exprt &arg=args[0];
+  irep_idt sval;
+
+  assert(arg.operands().size()==1);
+  if(arg.op0().operands().size()==2 &&
+     arg.op0().op0().id()==ID_string_constant)
+  {
+    // C string constant
+    const exprt &s=arg.op0().op0();
+    sval=to_string_constant(s).get_value();
+  }
+  else
+  {
+    // Java string constant
+    assert(refined_string_typet::is_unrefined_string_type(arg.type()));
+    const exprt &s=arg.op0();
+
+    // It seems the value of the string is lost,
```

Comment:
Always in the current state of CBMC. I have a branch a cbmc-testgen with some commits of Chris that improve this but I don't think there as been a pull request for it yet.